### ISSUE TITLE
Documentation is Incomplete

### ISSUE THREAD

**Issue by cjmorelock** [SharePoint/sp-dev-docs#5544]
Why don't you show explicitly where to put the &lt;AddContentTypeField&gt; element?  Does it go inside of the "upgrade action file", just like the &lt;ApplyElementManifests&gt; element is in this file?  Or does it go in an element manifest file, like where you add new &lt;Field&gt; and &lt;ContentType&gt; elements?  Also, why don't you give any information or example about how to use the VersionRange that is mentioned?  This is always how MS docs are. I'm not sure I've ever seen one that gives people enough information to actually use the tools that are discussed.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b4f8d79e-59f8-967b-3339-479c52be24c3
* Version Independent ID: 0bf6bc17-ac75-966a-cf48-69e1f6ced047
* Content: [Provision SharePoint assets with your solution package](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/toolchain/provision-sharepoint-assets#feedback)
* Content Source: [docs/spfx/toolchain/provision-sharepoint-assets.md](https://github.com/SharePoint/sp-dev-docs/blob/master/docs/spfx/toolchain/provision-sharepoint-assets.md)
* Product: **sharepoint**
* Technology: **sharepoint-framework**
* GitHub Login: @spdevdocs
* Microsoft Alias: **spdevdocs**

**Comment by bcameron1231** [comment]
Hi, thank you for the comment. I can understand that often times the documentation can be hard to follow. In many cases, the examples for showing how things are implemented are actually linked to another area of documentation.

In this case, you would want to follow the link for "[UpgradeActions](https://msdn.microsoft.com/library/office/ee537575(v=office.14).aspx)" in the [Upgrade Section](https://docs.microsoft.com/en-us/sharepoint/dev/spfx/toolchain/provision-sharepoint-assets#upgrade-sharepoint-items) of the documentation you provided. The link provided shows an example of how to use the VersionRange, and specifies AddContentTypeFiel> as a sub-element of the UpgradeActions node of your Feature.xml

Hope this helps.

**Comment by cjmorelock** [comment]
Thanks for the kind reply and the pointers.  So, in the package-solution.json file in the SPFx solution, there is a

"upgradeActions": []

property.  My guess is that this “upgradeActions” property, where you can specify one or more .xml files that contain your upgrade directives, replaces the <UpgradeActions> xml element that is shown in the OLD UpgradeActions documentation that you pointed me to.  So you would put your <AddContentTypeField>, <ApplyElementManifests> and <VersionRange> elements, as needed, in the .xml file(s) listed in the “upgradeActions” of the feature defined in the package-solution.json file, and you would NOT put them inside of a <UpgradeActions> parent element.  That is, in one of these .xml files you could have, for example, exactly and only this:

<ApplyElementManifests>
      <ElementManifest Location="9c0be970-a4d7-41bb-be21-4a683586db18\elements-v2.xml" />
</ApplyElementManifests>
  <AddContentTypeField
     ContentTypeId="0x010100A6F9CE1AFE2A48f0A3E6CB5BB770B0F7"
     FieldId="{B250DCFD-9310-4e2d-85F2-BE2DA37A57D2}"
     PushDown="TRUE" />
<VersionRange
    BeginVersion="2.0.0.0"
    EndVersion="5.0.0.0"></VersionRange>


and you would not have this:

<UpgradeActions>
<ApplyElementManifests>
      <ElementManifest Location="9c0be970-a4d7-41bb-be21-4a683586db18\elements-v2.xml" />
</ApplyElementManifests>
  <AddContentTypeField
     ContentTypeId="0x010100A6F9CE1AFE2A48f0A3E6CB5BB770B0F7"
     FieldId="{B250DCFD-9310-4e2d-85F2-BE2DA37A57D2}"
     PushDown="TRUE" />
<VersionRange
    BeginVersion="2.0.0.0"
    EndVersion="5.0.0.0"></VersionRange>
</UpgradeActions>

Can you confirm that this is true?

Thank you.

**Comment by bcameron1231** [comment]
Hi, yup exactly right. 

Have a look at this tutorial, it walks you through setting up the elements and upgrade actions. 

https://docs.microsoft.com/en-us/sharepoint/dev/spfx/web-parts/get-started/provision-sp-assets-from-package#define-upgrade-actions-for-new-version

**Comment by cjmorelock** [comment]
Thank you for the confirmation.  I appreciate your attention to this and your assistance.

I had already previously gone through this tutorial.  One of the problems is that the Microsoft documentation for SharePoint development is such a tangled mess that it is, in my view, usually impossible to know exactly what functionality is supported and what is not.  For example, this tutorial section says:

“
SharePoint Framework solutions do support the following Feature Framework upgrade action definitions:
·         ApplyElementManifest
·         AddContentTypeField
”

There is no mention of the <VersionRange> directive that you say that I can use.
